src/StaticLS/HIE/File.hs

```diff
@@ -0,0 +1,173 @@
+module StaticLS.HIE.File where
+
+import Control.Exception
+import Control.Monad
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.IO.Unlift
+import Control.Monad.Trans (lift)
+import Control.Monad.Trans.Maybe (MaybeT (..), runMaybeT)
+import Control.Monad.Trans.Reader (ReaderT (..), ask)
+import Data.Either (fromRight)
+import Data.List
+import Data.List.Extra (dropPrefix)
+import qualified Data.Map as Map
+import Data.Maybe (catMaybes, mapMaybe)
+import qualified Data.Set as Set
+import qualified Data.Text as T
+import qualified GHC
+import GHC.Data.Maybe (firstJusts)
+import qualified GHC.Iface.Ext.Binary as GHC
+import qualified GHC.Iface.Ext.Types as GHC
+import qualified GHC.Types.Name.Cache as GHC
+import qualified GHC.Unit.Types as GHC
+import HieDb
+import qualified Language.LSP.Types as LSP
+import StaticLS.StaticEnv
+import System.Directory (doesFileExist, makeAbsolute)
+import System.FilePath (normalise, (-<.>), (</>))
+
+type SrcFilePath = FilePath
+type HieFilePath = FilePath
+
+-- | Retrieve a hie info from a lsp text document identifier
+getHieFileFromTdi :: (HasStaticEnv m, MonadIO m) => LSP.TextDocumentIdentifier -> m (Maybe GHC.HieFile)
+getHieFileFromTdi tdi = do
+    staticEnv <- getStaticEnv
+    runMaybeT $ do
+        srcFilePath <- MaybeT $ pure $ LSP.uriToFilePath tdi._uri
+        hieFilePath <- MaybeT $ srcFilePathToHieFilePath srcFilePath
+        MaybeT $ getHieFile hieFilePath
+
+getHieFile :: (HasStaticEnv m, MonadIO m) => HieFilePath -> m (Maybe GHC.HieFile)
+getHieFile hieFilePath = do
+    staticEnv <- getStaticEnv
+    liftIO $
+        (Just <$> (pure . GHC.hie_file_result <=< GHC.readHieFile staticEnv.nameCache) hieFilePath)
+            -- Return nothing if the file failed to read
+            `catch` (\e -> let _ = (e :: IOException) in pure Nothing)
+
+modToHieFile :: (HasStaticEnv m, MonadIO m) => GHC.ModuleName -> m (Maybe GHC.HieFile)
+modToHieFile = runMaybeT . (MaybeT . getHieFile <=< MaybeT . modToHieFilePath)
```

Comment:
have you considered leaving all of these utilities in `MaybeT` so that you don't need to wrap / unwrap `MaybeT` everywhere?

in other words:

```haskell
modToHieFile :: (HasStaticEnv m, MonadIO m) => GHC.ModuleName -> MaybeT m GHC.HieFile
```

Reply:
Hm I think based on your above comment I'm going to try to convert these all to `ExceptT` in order to preserve the exception info as well.

Reply:
So I did some refactoring to use `ExceptT` in some places and `MaybeT` in others - I do think it's natural to "catch" the error and discard the information for the src -> hie file and vice versa functions but I try to return `ExceptT` from the others